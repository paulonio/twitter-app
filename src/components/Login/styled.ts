import { styled } from 'styled-components';

export const Form = styled.form`
  width: 100%;
  max-width: 450px;
  margin: 0 auto 40px;
  display: flex;
  flex-direction: column;
  align-items: start;

  @media screen and (max-width: 768px) {
    padding: 0;
  }
`;

export const LoginWrapper = styled.div`
  width: 33%;
  max-width: 450px;
  margin: 0 auto;
  display: flex;
  flex-direction: column;
  align-items: end;

  @media screen and (max-width: 1480px) {
    width: 50%;
  }

  @media screen and (max-width: 768px) {
    width: 75%;
  }

  @media screen and (max-width: 470px) {
    width: 100%;
    height: 100vh;
    padding: 0 8px;
    justify-content: center;
  }
`;
