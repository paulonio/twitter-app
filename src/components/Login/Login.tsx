import React from 'react';
import { useNavigate } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import { SubmitHandler, useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import * as yup from 'yup';
import { Form } from './styled';
import { WelcomeSubtitle } from '../Welcome/styled';
import { Error, Input } from '../SignUp/styled';
import Button from '../Button/Button';
import { SecondaryButton } from '../SidebarUsers/styled';
import { loginRequest } from '../../store/saga/authSaga';

export interface LoginForm {
  email: string;
  password: string;
}

const schema = yup
  .object({
    email: yup.string().email().required(),
    password: yup.string().min(6).max(20).required(),
  })
  .required();

const Login = () => {
  const navigate = useNavigate();
  const dispatch = useDispatch();

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<LoginForm>({
    resolver: yupResolver(schema),
  });

  const onSubmit: SubmitHandler<LoginForm> = (data) => {
    dispatch(loginRequest(data));
  };

  const handleNavigateToWelcomePage = () => {
    navigate('/');
  };

  return (
    <>
      <Form onSubmit={handleSubmit(onSubmit)}>
        <WelcomeSubtitle>Log in to Twitter</WelcomeSubtitle>
        <Input type="text" placeholder="Email" {...register('email')} />
        <Error>{errors.email?.message}</Error>
        <Input type="password" placeholder="Password" {...register('password')} />
        <Error>{errors.password?.message}</Error>
        <Button buttonType="primary">Log In</Button>
      </Form>
      <SecondaryButton onClick={handleNavigateToWelcomePage}>Sign up to Twitter</SecondaryButton>
    </>
  );
};

export default Login;
