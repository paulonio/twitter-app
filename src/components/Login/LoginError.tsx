import React, { FC } from 'react';
import { useDispatch } from 'react-redux';
import { clearError } from '@src/store/slices/authSlice';
import Modal from '@components/Modal/Modal';

import { Error } from './styled';

interface LoginErrorProps {
  error: string;
}

const LoginError: FC<LoginErrorProps> = ({ error }) => {
  const dispatch = useDispatch();

  const handleClose = () => {
    dispatch(clearError());
  };

  return (
    <Modal isActive={Boolean(error)} setActive={handleClose}>
      <Error>{error.split('-').join(' ')}</Error>
    </Modal>
  );
};

export default LoginError;
