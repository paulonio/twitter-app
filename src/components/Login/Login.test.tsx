import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

import Login from '@components/Login/Login';
import CustomProvider from '../EditUser/testUtils';

describe('Login', () => {
  it('should render login component', () => {
    const { container } = render(
      <CustomProvider>
        <Login />
      </CustomProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should change login inputs', () => {
    render(
      <CustomProvider>
        <Login />
      </CustomProvider>
    );

    const emailInput: HTMLInputElement = screen.getByTestId('email');
    const passwordInput: HTMLInputElement = screen.getByTestId('password');
    fireEvent.change(emailInput, { target: { value: 'Test email' } });
    fireEvent.change(passwordInput, { target: { value: 'Test password' } });
    expect(emailInput.value).toEqual('Test email');
    expect(passwordInput.value).toEqual('Test password');
  });

  it('should set errors on incorrect values', async () => {
    render(
      <CustomProvider>
        <Login />
      </CustomProvider>
    );

    const emailInput: HTMLInputElement = screen.getByTestId('email');
    const passwordInput: HTMLInputElement = screen.getByTestId('password');
    const submit: HTMLButtonElement = screen.getByTestId('submit-button');
    fireEvent.change(emailInput, { target: { value: 'Test' } });
    fireEvent.change(passwordInput, { target: { value: 'Test' } });
    fireEvent.click(submit);
    await waitFor(() => {
      expect(screen.getByText('email must be a valid email')).toBeInTheDocument();
      expect(screen.getByText('password must be at least 6 characters')).toBeInTheDocument();
    });
  });
});
