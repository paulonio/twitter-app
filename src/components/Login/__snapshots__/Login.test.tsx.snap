// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render login component 1`] = `
<div>
  <div
    class="sc-bGbknN bMWdwO"
  >
    <img
      alt="Twitter logo"
      class="sc-hlwzUd egxnlX"
      src="/assets/icons/twitter-logo.svg"
    />
    <form
      class="sc-kIyXsW ceryGg"
    >
      <h2
        class="sc-jxAitF dxAOsA"
      >
        Log in to Twitter
      </h2>
      <input
        class="sc-hzJlrQ jUPwRT"
        data-testid="email"
        name="email"
        placeholder="Email"
        type="text"
      />
      <p
        class="sc-hlCNGy kUiYVS"
      />
      <input
        class="sc-hzJlrQ jUPwRT"
        data-testid="password"
        name="password"
        placeholder="Password"
        type="password"
      />
      <p
        class="sc-hlCNGy kUiYVS"
      />
      <button
        class="sc-cEPTvB eFzckz"
        data-testid="submit-button"
      >
        Log In
      </button>
    </form>
    <button
      class="sc-cEPTvB cdTOtA"
    >
      Sign up to Twitter
    </button>
  </div>
</div>
`;
